NitroxClient/GameLogic/InitialSync/BuildingInitialSyncProcessor.cs

```diff
@@ -74,5 +92,97 @@ private void FinishedCompletedBuildings(object sender, EventArgs eventArgs)
             ThrottledBuilder.main.QueueDrained -= FinishedCompletedBuildings;
             completed = true;
         }
+
+        // Subnautica changes the Layout of Bases when new BasePieces are added or removed. A simple replay of all the saved BasePieces is not enough, because 
+        // deconstructig BasePieces (like Tubes) and building new BasePieces can cause the complete layout to change, so that following Pieces 
+        // (like Hatches) aren't fit right to the corresponding objects. Additionally we need to change the 
+        // The Method is not final. Some BasePiece Relations may be better validated by Face- or Cell-Connections, others still need fixed positons like 
+        // Solarpanels or Furniture. For now the BuildIndex should work. 
+        private List<BasePiece> GenerateBaseLayouts(List<BasePiece> basePieces)
```

Comment:
Can we refactor this to have an ordered array of base piece types and cross reference that?  I think such an approach will remove the need to have so many loops.

Reply:
Thinking more about this total method makes me unsure, if this can be eventually mostly dropped and only the buildindex and base sequences are needed. 
The first problem were unfinished peaces. These unfinished pieces should be now correctly handled, even if they are not loaded at the end of a base layout. 
The second problem is hull integrity. This will be overcome as soon as i found out how to suppress this on initial sync. My idea herefore is, to sort all items via parentbase and suppress hull integrity calculation except for the last finished piece of this base to let the game refresh the latest state. (For a more later PR: Thinking a little bit more further, the hull integrity sync could also be done in another separate InitialSync after the BuildingInitialsync which also can handle breaches and floot level of the base).
The third problem was energy. This should also be fixed, because alle energy producing and energy consuming objects are now correctly referenced by the corresponding base object and even if the energy generating object has been placed afterwards it should be just working as in normal singleplayer. 
The fourth problem were hatches on watertanks and on rooms for creating new corridors, this should also be no problem if the buildindex is correct. 

For now i would like to let the hullintegrity related sorts as they are and comment the code with the infos given here. 

For the use of an arry and cross referencing, can you give me a hint on that? I don't know exactly how you are meaning this.

Reply:
Implemented also sorting via ParentID in the initial sorting. 

The rest of the item adding i would let as it is, because this is only executed once in a game session and this code is easy to understand an readable.